File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.py
def solve(depth,n):
    global result
    if depth == N//2:
        start,link = 0,0
        for i in range(N):
            for j in range(N):
                if visited[i] and visited[j]:
                    start += S[i][j]
                elif not visited[i] and not visited[j]:
                    link += S[i][j]
        result= min(result, abs(start-link))
        return result
    else:
        for i in range(n,N):
            if not visited[i]:
                visited[i] = True
                solve(depth+1,i+1)
                visited[i] = False
        
N = int(input())
S = [list(map(int,input().split())) for i in range(N)]
visited=[0]*N
result = float('inf')
solve(0,0)
print(result)

File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.cc
#include <vector>
#include <algorithm>
#include <iostream>
#include <climits>
#include <set>
using namespace std;
typedef long long ll;

int N, result = INT_MAX;
vector<vector<int>> S;
vector<int> member;
vector<int> teamA;
vector<int> teamB;
// 팀을 다 꾸리기 N/2

int getPower(vector<int> team) {
	vector<pair<int, int>> pairs;
	
	for (int i = 0; i < team.size(); i++) {
		for (int j = i + 1; j < team.size(); j++) {
			pairs.push_back({ team[i],team[j] });
		}
	}
	int power = 0;
	for (pair<int, int> p : pairs) {
		power += (S[p.first][p.second] + S[p.second][p.first]);
	}
	return power;
}
void backtraking(int index,int depth) {
	
	if (depth == N / 2 ) {
	
		teamB.resize(N/2);
		auto it = set_difference(member.begin(), member.end(), teamA.begin(), teamA.end(), teamB.begin());
		//teamB.erase(it, teamB.end()); //남은공간 제거 코드
		int A = getPower(teamA);
		int B = getPower(teamB);
		result = min(result, abs(A - B));
		return;
	}
	
	for (int i = index; i < N; i++) {
		teamA.push_back(i);
		backtraking(i+1,depth + 1);
		teamA.pop_back();
	}
}
int main() {
	cin >> N; S.assign(N, vector<int>(N));

	for (int i = 0; i < N; i++)
		for (int j = 0; j < N; j++) cin >> S[i][j];

	for (int i = 0; i < N; i++) {
		member.push_back(i);
	}
	backtraking(0,0);

	cout << result;
}
